File: PAT_Basic_Level_by_r/1077 互评成绩计算/1077 互评成绩计算.R
# 题目名称：互评成绩计算
# 题目编号：PAT Basic 1077
# 实现原理：
#   对每个组的评分进行处理：将所有分数排序后去掉一个最高分和一个最低分，
#   计算剩余分数的平均值，再与老师评分（注：此处代码逻辑为取排序后去首尾的均值），
#   最后四舍五入取整输出最终得分。
# 时间复杂度：O(N * N log N)，N为分组数，每组排序需要O(N log N)
# 空间复杂度：O(N)，存储一组分数
x<-readLines("stdin",warn=FALSE);h<-as.integer(strsplit(x[1],"\\s+")[[1]]);for(s in x[2:(h[1]+1)]){v<-as.numeric(strsplit(s,"\\s+")[[1]]);teacher<-v[1];peer<-v[-1];peer<-peer[peer>=0&peer<=h[2]];peer<-sort(peer);peer<-peer[-c(1,length(peer))];ans<-floor((mean(peer)+teacher)/2+0.5);cat(ans,"\n")}

File: PAT_Basic_Level_by_lua/1077 互评成绩计算/1077 互评成绩计算.lua
-- 题目编号: 1077 互评成绩计算
-- 实现原理: 对每个学生，读取教师评分和一组互评分，剔除无效分数和最高最低分后取平均，与教师评分取均值并四舍五入输出。
local n, m = io.read("*n"), io.read("*n")  -- n为学生数，m为互评分数量

-- 遍历每个学生
for i = 1, n do
    local teacher = io.read("*n")  -- 读取教师评分
    local a = {}  -- 存储有效的互评分
    -- 读取其余 n-1 个互评分
    for j = 1, n - 1 do
        local x = io.read("*n")  -- 读取一个互评分
        -- 筛选有效分数：在0到满分之间
        if x >= 0 and x <= m then
            a[#a + 1] = x
        end
    end
    table.sort(a)  -- 对有效互评分排序，便于去除最高分和最低分
    local s = 0  -- 累加求和变量
    -- 去掉一个最高分和一个最低分后求和（从第2个到倒数第2个）
    for j = 2, #a - 1 do
        s = s + a[j]
    end
    -- 计算最终成绩：教师分与互评分均值各占50%，四舍五入
    -- math.floor(x + 0.5) 实现四舍五入
    print(math.floor((teacher + s / (#a - 2)) / 2 + .5))
end
